### ISSUE TITLE
[Kindle NT] Full screen refresh doesn't work

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#12774]
* KOReader version: 2024.07-285-gfd3727274 (legacy)
* Device: Kindle 4 Non-Touch (FW 4.1.4)

#### Issue
Full screen refresh doesn't work, as reported here: https://github.com/koreader/koreader/pull/12484#issuecomment-2422787161

#### Steps to reproduce
Assign a hotkey to refresh the screen.
Use it to refresh - nothing happens.

Well, I found that it **_sometimes_** works, but cannot trace the lead.
See `crash.log` below.

##### `crash.log` (if applicable)
[crash_partial.log.txt](https://github.com/user-attachments/files/17851923/crash_partial.log.txt)
In this one, I did:
1) Use hotkey to refresh (1) - nothing
2) Jump back to previous chapter using Dpad left.
3) Use hotkey to refresh (2) - nothing
4) Use hotkey to refresh **again** (3) - it refreshed!
5) (outside `crash.log` Use hotkey to refresh n more time - nothing.

Although, logs looks exactly the same for all three refreshes above.
Need some more debugging...

**Comment by mergen3107** [comment]
This is how it looks like when I have `Refresh on chapter boundaries on` and skip to next chapter:
```
11/21/24-16:23:16 DEBUG CreDocument: goto page 431 flow 0 
11/21/24-16:23:16 DEBUG triggering refresh {
  mode = "full",
  region = "600x800+0+0"
} --[[table: 0x44cf4a28]] 
11/21/24-16:23:16 DEBUG key event => code: 106 (Right), value: 0, time: 1732224196.345777 
11/21/24-16:23:16 DEBUG AutoSuspend: onInputEvent 
```

**Comment by mergen3107** [comment]
@Commodore64user 
Unrelated to that issue, I see this when I open a menu:
```
11/21/24-16:22:45 WARN  menu id not found: taps_and_gestures 
```

**Comment by mergen3107** [comment]
So, if I keep mashing the screen refresh combo, it eventually refreshes once, but then never again.

I wonder if there are any timing clashes.

I tried disabling `AutoSuspend`, but no avail.

**Comment by mergen3107** [comment]
Haha, weird! Look :D 
In ssh on Kindle, I did `eips -c` to clear the screen.
Screen went blank (OK).
I hit the combo to refresh again - and I only got status bar to refresh!!

So, I think, status bar somehow is stealing the full screen refresh show...

**Comment by mergen3107** [comment]
![Reader_2024-11-21_164534](https://github.com/user-attachments/assets/96e734d0-b9c0-48fd-b106-4f2d61610ee1)

**Comment by mergen3107** [comment]
Et voilà!
I found the actual culprit.
Disable `Hide battery item at custom threshold`...
@hius07

**Comment by mergen3107** [comment]
Now, every single combo to refresh actually causes the refresh!

**Comment by Commodore64user** [comment]
> @Commodore64user Unrelated to that issue, I see this when I open a menu:
> 
> ```
> 11/21/24-16:22:45 WARN  menu id not found: taps_and_gestures 
> ```

Unimportant as that is the menu for gestures (which we don’t need), there is a device check for that and we are not a touch device ;) Nothing to worry about.

**Comment by Commodore64user** [comment]
> Et voilà! I found the actual culprit. Disable `Hide battery item at custom threshold`... @hius07

My man here, doing the lord’s work… nice one ☝️

**Comment by mergen3107** [comment]
You're too kind! :D 
I am just poking around with everything I know so far xD

**Comment by mergen3107** [comment]
> Unimportant as that is the menu for gestures (which we don’t need)

Ah, yes, I forgot (again).
In this case, we could possible add it to this check (if needed to fix at all :D)
```
or (not Device:isTouchDevice() and "taps_and_gestures")
```

https://github.com/koreader/koreader/blob/346512d80cbf36253ac4b41fe2df15c9e12dc696/frontend/ui/menusorter.lua#L102-L105

**Comment by mergen3107** [comment]
Interesting side observation.
While reading a book, hitting `ScreenKB+Press` (which is "right click" in most cases) also refreshes the screen.

**Comment by Frenzie** [comment]
> In this case, we could possible add it to this check (if needed to fix at all :D)

I'd rather not, or at least not there.

**Comment by Commodore64user** [comment]
> Interesting side observation. While reading a book, hitting `ScreenKB+Press` (which is "right click" in most cases) also refreshes the screen.

yep, you set it up that way though https://github.com/koreader/koreader/pull/12484#issuecomment-2484724750

```
["modifier_plus_press"] = {
    ["full_refresh"] = true,
    ["settings"] = {
        ["order"] = {
            [1] = "full_refresh",
        },
    },
},
```

**Comment by mergen3107** [comment]
Here is the related battery threshold PR: https://github.com/koreader/koreader/pull/7876

**Comment by mergen3107** [comment]
My intuition tells me that it might related to the function that regularly checks for battery level to make decision about hiding the icon in the status bar.

**Comment by mergen3107** [comment]
and then, if I also uncheck `Hide inactive items`, refresh works again :D 
What

**Comment by mergen3107** [comment]
and then, if I also uncheck `Page turning inverted`, it stops working again...
wat

**Comment by mergen3107** [comment]
If I now enable battery icon again, then it works again.
I think there is some kind of combination of icons that kills refresh.

**Comment by mergen3107** [comment]
Added cropped images for better illustration, in case you guys want to reproduce it.

**Comment by mergen3107** [comment]
@NiLuJe 
Status bar is misbehaving again, can you please have a look at this issue xD

**Comment by NiLuJe** [comment]
Misbehaving how, exactly?

(And can you reproduce with a "from-scratch" config, and not an imported one? ;)).

**Comment by mergen3107** [comment]
See my last 3-4 posts above. 

Enabling/disabling items/features in status bar makes full refresh go away and come back. 

I will try on a fresh build, but at least @Commodore64user can confirm that full refresh doesn't work if `Hide battery item at custom threshold` is enabled.

**Comment by hius07** [comment]
I assigned Full refresh to a gesture (Voyage) and it works good.
Can you give exact combination of status bar settings when it doesn't work for you.

**Comment by mergen3107** [comment]
First thing that prevented full refresh was this one: https://github.com/koreader/koreader/issues/12774#issuecomment-2492411252

**Comment by hius07** [comment]
I've tried before asking, all good.
Looks like some combination of indicators/settings is bad.

**Comment by mergen3107** [comment]
That's interesting, that might be non-touch only issue then...

**Comment by mergen3107** [comment]
I don't know details, but I think I cannot find how to "tap on status bar to hide/unhide it" on K4NT.
Also, I remember there was a settings called `Lock status bar` that prevents that tap and status bar stay on despite taps.

@Commodore64user 
DO you know how/when this function was removed?

**Comment by hius07** [comment]
Actually the footer can somehow affect on it, the only handler is
https://github.com/koreader/koreader/blob/346512d80cbf36253ac4b41fe2df15c9e12dc696/frontend/device/devicelistener.lua#L424-L429

**Comment by hius07** [comment]
As @NiLuJe said we now tend to use `maybeUpdateFooter`.

**Comment by hius07** [comment]
Can you try
```Lua
if self.view then
    self.view.footer:maybeUpdateFooter()
end
```

**Comment by mergen3107** [comment]
In parallel, I am trying to do it on a clean `koreader` folder.
I cannot disable `Hide battery` for status bar :D 
By default, it is enabled, and threshold is 100%. When I try `Disable`, it doesn't get disbaled, and threshold go up by 1, so it becomes 101, 102, etc, once I try disable Oo.

**Comment by hius07** [comment]
If you comment out updating the footer, all works, right?

**Comment by mergen3107** [comment]
I enabled verbose logging and restarted, now I can disable. I'll try again later to see if it was just a restart that fixed it, or verbose logging too.

**Comment by mergen3107** [comment]
> If you comment out updating the footer, all works, right?

@hius07 
Which line is this please?

**Comment by hius07** [comment]
maybeUpdateFooter

**Comment by mergen3107** [comment]
@NiLuJe 
It looks similar on a fresh `koreader` build too.
Initially it doesn't work, and I need to enable `Hide inactive items` and disable `Hide battery`.
Then it breaks again after I uncheck `Battery indicator`...

**Comment by hius07** [comment]
I'm talking about the refresh only.

**Comment by mergen3107** [comment]
Yeah, sorry, I was doing clean tests in between.

Back to my current build, I commented it out:
```
function DeviceListener:onFullRefresh()
    -- if self.ui and self.ui.view then
        -- self.ui:handleEvent(Event:new("UpdateFooter", self.ui.view.footer_visible))
    -- end
    if self.view then
        -- self.view.footer:maybeUpdateFooter()
    end
    UIManager:setDirty(nil, "full")
end
```
but it still doesn't refresh on those initial settings...

**Comment by mergen3107** [comment]
It still thinks it is doing it:
```
11/22/24-12:06:11 DEBUG Executing actions for hotkey:  modifier_plus_press  with events:  {
  full_refresh = true,
  settings = {
    order = {
      "full_refresh"
    } --[[table: 0x428b2c48]]
  } --[[table: 0x416c9258]]
} --[[table: 0x4058b3d8]] 
11/22/24-12:06:11 DEBUG triggering refresh {
  mode = "full",
  region = "600x800+0+0"
} --[[table: 0x41728c78]] 
```

**Comment by hius07** [comment]
The footer is not involved at all.

**Comment by mergen3107** [comment]
OK, so with the footer refresh commented out, full refresh doesn't work in any of the combinations of the status bar.

**Comment by mergen3107** [comment]
Yep, nothing refreshes after `eips -c` and full refresh hotkey.

**Comment by mergen3107** [comment]
Something happens after this event fires:
```
11/22/24-13:05:30 DEBUG triggering refresh {
  mode = "full",
  region = "600x800+0+0"
} --[[table: 0x477c1cb8]] 
```
This is right after the hotkey:
```
11/22/24-13:05:30 DEBUG Executing actions for hotkey:  modifier_plus_press  with events:  {
  full_refresh = true,
  settings = {
    order = {
      "full_refresh"
    } --[[table: 0x428b2c48]]
  } --[[table: 0x416c9258]]
} --[[table: 0x4058b3d8]] 
```

**Comment by mergen3107** [comment]
This is `_repaint` function of the `UIManager`

**Comment by Commodore64user** [comment]
> I don't know details, but I think I cannot find how to "tap on status bar to hide/unhide it" on K4NT.
> 
> Also, I remember there was a settings called `Lock status bar` that prevents that tap and status bar stay on despite taps.
> 
> 
> 
> @Commodore64user 
> 
> DO you know how/when this function was removed? 

You need to start thinking buttons and stop thinking touch screens ;).

We simply can't tap the screen to hide/show the status bar. It [that setting] sits behind a device check however, you can now technically do it with shortcuts... there is a dispatcher event to toggle it.

**Comment by mergen3107** [comment]
> It [that setting] sits behind a device check 

OK :D No, I was just looking around to try to connect to refresh.

I saw the thing, looks neat to be able to toggle it with a shortcut

**Comment by mergen3107** [comment]
I setup `LOGG 2` logs to show up exactly after device listener call:
DeviceListener:
```
function DeviceListener:onFullRefresh()
    -- if self.ui and self.ui.view then
        -- self.ui:handleEvent(Event:new("UpdateFooter", self.ui.view.footer_visible))
    -- end
    -- if self.view then
        -- self.view.footer:maybeUpdateFooter()
    -- end
    UIManager:setDirty(nil, "full")
    logger.dbg("LOGG 1 DeviceListener:onFullRefresh TRIGGERED")
end
```

UIManager:
```
function UIManager:setDirty(widget, refreshtype, refreshregion, refreshdither)
    logger.dbg("LOGG 2 widget = ", widget)
    logger.dbg("LOGG 2 refreshtype = ", refreshtype)
    logger.dbg("LOGG 2 refreshregion = ", refreshregion)
    logger.dbg("LOGG 2 refreshdither = ", refreshdither)
```

**Comment by mergen3107** [comment]
`LOGG 1` shows up, although before the button combo (?)
but then `LOGG 2` never shows up.

**Comment by mergen3107** [comment]
I waited a while and then hit combo again. Here is everything to goes on:
```
11/22/24-17:23:07 DEBUG input event => type: 4 (EV_MSC), code: 4 (nil), value: 0, time: 1732314187.089654 
11/22/24-17:23:07 DEBUG key event => code: 29 (ScreenKB), value: 1, time: 1732314187.089720 
11/22/24-17:23:07 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 1732314187.090162 
11/22/24-17:23:07 DEBUG key event => code: 194 (Press), value: 1, time: 1732314187.216883 
11/22/24-17:23:07 DEBUG AutoSuspend: onInputEvent 
11/22/24-17:23:07 DEBUG LOGG 2 widget =  nil 
11/22/24-17:23:07 DEBUG LOGG 2 refreshtype =  full 
11/22/24-17:23:07 DEBUG LOGG 2 refreshregion =  nil 
11/22/24-17:23:07 DEBUG LOGG 2 refreshdither =  nil 
11/22/24-17:23:07 DEBUG LOGG 3 refreshdither nil 
11/22/24-17:23:07 DEBUG LOGG 4 refreshdither nil 
11/22/24-17:23:07 DEBUG _refresh: Enqueued full update for region 0 0 600 800 dithering: false 
11/22/24-17:23:07 DEBUG setDirty full from widget nil w/ NO region; dithering: nil 
11/22/24-17:23:07 DEBUG LOGG 1 DeviceListener:onFullRefresh TRIGGERED 
11/22/24-17:23:07 DEBUG Executing actions for hotkey:  modifier_plus_press  with events:  {
  full_refresh = true,
  settings = {
    order = {
      "full_refresh"
    } --[[table: 0x417e4028]]
  } --[[table: 0x417e3fc8]]
} --[[table: 0x417e3f68]] 
11/22/24-17:23:07 DEBUG triggering refresh {
  mode = "full",
  region = "600x800+0+0"
} --[[table: 0x430c2390]] 
11/22/24-17:23:07 DEBUG LOGG X refresh.region.x 0 
11/22/24-17:23:07 DEBUG LOGG X refresh.region.y 0 
11/22/24-17:23:07 DEBUG LOGG X refresh.region.w 600 
11/22/24-17:23:07 DEBUG LOGG X refresh.region.h 800 
11/22/24-17:23:07 DEBUG key event => code: 194 (Press), value: 0, time: 1732314187.490589 
11/22/24-17:23:07 DEBUG AutoSuspend: onInputEvent 
11/22/24-17:23:07 DEBUG input event => type: 4 (EV_MSC), code: 4 (nil), value: 0, time: 1732314187.499598 
11/22/24-17:23:07 DEBUG key event => code: 29 (ScreenKB), value: 0, time: 1732314187.499614 
11/22/24-17:23:07 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 1732314187.499617 
```

**Comment by mergen3107** [comment]
And this is how it looks like when it works (with these settings https://github.com/koreader/koreader/issues/12774#issuecomment-2492531864)

I enabled back the status bar refresh:
```
function DeviceListener:onFullRefresh()
    if self.ui and self.ui.view then
        self.ui:handleEvent(Event:new("UpdateFooter", self.ui.view.footer_visible))
    end
    -- if self.view then
        -- self.view.footer:maybeUpdateFooter()
    -- end
    UIManager:setDirty(nil, "full")
    logger.dbg("LOGG 1 DeviceListener:onFullRefresh TRIGGERED")
end
```
Logs:
```
11/22/24-17:29:36 DEBUG input event => type: 4 (EV_MSC), code: 4 (nil), value: 0, time: 1732314576.403142 
11/22/24-17:29:36 DEBUG key event => code: 29 (ScreenKB), value: 1, time: 1732314576.403205 
11/22/24-17:29:36 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 1732314576.403651 
11/22/24-17:29:36 DEBUG key event => code: 194 (Press), value: 1, time: 1732314576.641772 
11/22/24-17:29:36 DEBUG Explicit widgetRepaint: table: 0x412207e8 @ 0 0 
11/22/24-17:29:36 DEBUG LOGG 2 widget =  nil 
11/22/24-17:29:36 DEBUG LOGG 2 refreshtype =  function: 0x412375d8 
11/22/24-17:29:36 DEBUG LOGG 2 refreshregion =  nil 
11/22/24-17:29:36 DEBUG LOGG 2 refreshdither =  nil 
11/22/24-17:29:36 DEBUG LOGG 3 refreshdither nil 
11/22/24-17:29:36 DEBUG LOGG 4 refreshdither nil 
11/22/24-17:29:36 DEBUG setDirty via a func from widget nil 
11/22/24-17:29:36 DEBUG LOGG 2 widget =  nil 
11/22/24-17:29:36 DEBUG LOGG 2 refreshtype =  full 
11/22/24-17:29:36 DEBUG LOGG 2 refreshregion =  nil 
11/22/24-17:29:36 DEBUG LOGG 2 refreshdither =  nil 
11/22/24-17:29:36 DEBUG LOGG 3 refreshdither nil 
11/22/24-17:29:36 DEBUG LOGG 4 refreshdither nil 
11/22/24-17:29:36 DEBUG _refresh: Enqueued full update for region 0 0 600 800 dithering: false 
11/22/24-17:29:36 DEBUG setDirty full from widget nil w/ NO region; dithering: nil 
11/22/24-17:29:36 DEBUG LOGG 1 DeviceListener:onFullRefresh TRIGGERED 
11/22/24-17:29:36 DEBUG Executing actions for hotkey:  modifier_plus_press  with events:  {
  full_refresh = true,
  settings = {
    order = {
      "full_refresh"
    } --[[table: 0x42dac860]]
  } --[[table: 0x42dac800]]
} --[[table: 0x4181c2b0]] 
11/22/24-17:29:36 DEBUG _refresh: Enqueued full update for region 0 0 600 800 dithering: false 
11/22/24-17:29:36 DEBUG triggering refresh {
  mode = "full",
  region = "600x800+0+0"
} --[[table: 0x42d7e850]] 
11/22/24-17:29:36 DEBUG LOGG X refresh.region.x 0 
11/22/24-17:29:36 DEBUG LOGG X refresh.region.y 0 
11/22/24-17:29:36 DEBUG LOGG X refresh.region.w 600 
11/22/24-17:29:36 DEBUG LOGG X refresh.region.h 800 
11/22/24-17:29:36 DEBUG key event => code: 194 (Press), value: 0, time: 1732314576.852434 
11/22/24-17:29:37 DEBUG input event => type: 4 (EV_MSC), code: 4 (nil), value: 0, time: 1732314577.009547 
11/22/24-17:29:37 DEBUG key event => code: 29 (ScreenKB), value: 0, time: 1732314577.009562 
11/22/24-17:29:37 DEBUG input event => type: 0 (EV_SYN), code: 0 (SYN_REPORT), value: 0, time: 1732314577.009565 
```

**Comment by mergen3107** [comment]
I don't know what is different here.

**Comment by Commodore64user** [comment]
> `LOGG 1` shows up, although before the button combo (?) but then `LOGG 2` never shows up.

Makes sense, i placed the log *after* dispatching the event. Perhaps i should swap them around.

**Comment by NiLuJe** [comment]
The log looks sound, I'd trace what `onUpdateFooter` is doing instead.

**Comment by mergen3107** [comment]
@NiLuJe 
So it is normal that when I commented out UpdateFooter altogether, and only left setDirty function, it didn't refresh?

**Comment by NiLuJe** [comment]
> @NiLuJe So it is normal that when I commented out UpdateFooter altogether, and only left setDirty function, it didn't refresh?

Yes, `setDirty("nil", ...)` will only queue up a screen refresh (which we saw in the logs, IIRC), with no widget repaints of any kind.

**Comment by hius07** [comment]
But we can refresh the FM, without the footer.

**Comment by NiLuJe** [comment]
And to clarify, because I'm confused now:

- Are you talking about not seeing a screen refresh *at all*?

OR

- Are you talking about not seeing the footer being updated during said refresh?

(The K4 has a shadow buffer check that will optimize out refreshes when no content has been changed. I do not recall if this applies to flashing refreshes, though. i.e., there may be a very good reason why refreshes are optimized out that's entirely out of our hands ;)).

**Comment by hius07** [comment]
We are talking about the screen refresh only, not widgets uodate.

**Comment by Commodore64user** [comment]
> Are you talking about not seeing a screen refresh at all

this is the problem

**Comment by NiLuJe** [comment]
Then as long as you see

```
11/22/24-17:23:07 DEBUG triggering refresh {
  mode = "full",
  region = "600x800+0+0"
} --[[table: 0x430c2390]] 
```

(which i can see in your log), everything's going according to plan, and the refresh was simply dropped by the driver (see above).

You can add an `fb.debug` print in `ffi/framebuffer_einkfb.lua` (much like we do in `ffi/framebuffer_mxcfb.lua` (the `mxc_update` one) to convince yourself of that ;).

(I also don't remember if the driver drops it entirely silently or not, so you might want to check `dmesg` in case it whines about it in there).

**Comment by Commodore64user** [comment]
> The K4 has a shadow buffer check that will optimize out refreshes when no content has been changed. I do not recall if this applies to flashing refreshes, though. i.e., there may be a very good reason why refreshes are optimized out that's entirely out of our hands ;)).

You still can do manual refreshes to your heart’s content by using screnkb plus back with stock

**Comment by Commodore64user** [comment]
Get that k4 out and see it for yourself…

**Comment by mergen3107** [comment]
> Are you talking about not seeing a screen refresh at all?

Yes, I am talking about when you do "full screen refresh" hotkey and don't see ANY refresh anywhere.

**Comment by mergen3107** [comment]
No rush @NiLuJe :D 
We appreciate your help!

In the meantime, full screen refresh can be replaced by hitting menu button twice. It brings Top menu and moves it away, so it refreshes the whole screen. 

Or use one of the "lucky" status bar combos above (which I ended up with).

**Comment by mergen3107** [comment]
> Add to that the fact that the K4 doesn't actually use einkfb hardware, but is actually using an einkfb shim on top of mxcfb hardware, and things can get wonky pretty fast ;).

Taking into account that this is almost the only one issue I found with the screen so far, it is impressive

**Comment by mergen3107** [comment]
Following and rethinking steps I did starting from here https://github.com/koreader/koreader/issues/12774#issuecomment-2492519602,

I think refresh doesn't get triggered if status bar doesn't get "dirty" enough.

It is "dirty" and refresh works:
- If battery status is present
- If there are enough items in the status bar (unchecking "hide inactibe icons or also have "page turn reverse indicator)
- If time changed. This might explain why I was able to trigger refresh only "sometimes": because minute changed, status bar got dirty, and refresh queue was happy to refresh it.

I don't have K4NT next to me right now, but I would like to test more:
1) Disable status bar and try refreshes. I want to see if the issue is caused by things currently rendered on screen or due to status bar code hiccup. 
2) Move up "setDirty function" so it gets triggered before status bar's UpdateFooter function - in case it does something wrong and that's why I rarely saw setDirty logs as I reported above.

**Comment by mergen3107** [comment]
> You can add an fb.debug print

This?

https://github.com/koreader/koreader-base/blob/fccfa331bd24786a7010085629a2998ec4cf67e7/ffi/framebuffer_mxcfb.lua#L423

**Comment by NiLuJe** [comment]
> * dirty, and refresh queue was happy to refresh it

*We* don't care; the *driver* does.

(Note that this is still actually a thing in practice on newer devices (i.e., even GC, on its own, will not actually touch unchanged pixels), it's just that the APIs are now slightly more complex, and offer explicit semantics where things like GC + FULL will *always* refresh the screen, regardless of pixel status).

**Comment by NiLuJe** [comment]
> https://github.com/koreader/koreader-base/blob/fccfa331bd24786a7010085629a2998ec4cf67e7/ffi/framebuffer_mxcfb.lua#L423

Yup!

**Comment by mergen3107** [comment]
> _We_ don't care; the _driver_ does.

Well, OK :D To me (a non-dev) it is all the same "it doesn't work", but I got your point.

**Comment by mergen3107** [comment]
> i.e., even GC, on its own,

What is GC?

**Comment by mergen3107** [comment]
> You can add an fb.debug print in ffi/framebuffer_einkfb.lua (much like we do in ffi/framebuffer_mxcfb.lua (the mxc_update one) to convince yourself of that ;).

OK, I think I see your point. Both cases (working and non-working full refresh with a hotkey) are virtually identical:

[LOGG1_good_refresh_hotkey.txt](https://github.com/user-attachments/files/17891586/LOGG1_good_refresh_hotkey.txt)
[LOGG2_bad_refresh_hotkey.txt](https://github.com/user-attachments/files/17891587/LOGG2_bad_refresh_hotkey.txt)

> I also don't remember if the driver drops it entirely silently or not, so you might want to check dmesg in case it whines about it in there

It seems like it is completely silent, I don't think dmesg is useful at all.
It didn't change much before/after those hotkey full refresh actions.
Maybe you can see something in there @NiLuJe ?
[dmesg.txt](https://github.com/user-attachments/files/17891589/dmesg.txt)

**Comment by mergen3107** [comment]
What's next anyway?

Any chance we can get "force full refresh anyway no matter what"?

**Comment by mergen3107** [comment]
> 1. Disable status bar and try refreshes. I want to see if the issue is caused by things currently rendered on screen or due to status bar code hiccup.

So, I used the shortcut to disable status bar. I tried refreshes in a config where full refresh always works. After disabling status bar, it refreshes only the first time and that's it.

> * If time changed.

I wasn't able to confirm that. I used a config where it usually doesn't work.
Waited for a minute change. `Auto refresh status bar` is on, so I saw minute change in the status bar.
Tried hotkey refresh - nothing.

**Comment by mergen3107** [comment]
> 2\. Move up "setDirty function" so it gets triggered before status bar's UpdateFooter function - in case it does something wrong and that's why I rarely saw setDirty logs as I reported above.

Doesn't make any difference, reverting back.

**Comment by mergen3107** [comment]
> The K4 has a shadow buffer check that will optimize out refreshes when no content has been changed.

So, I found it weird. I put status bar in config where hotkey refresh doesn't usually work.
I tried refresh - nothing (as usual).
I advanced to the next page (content changed, right?). Tried refresh again - nothing again. (?)

**Comment by mergen3107** [comment]
Shouldn't it have refreshed this time?

**Comment by mergen3107** [comment]
> Any chance we can get "force full refresh anyway no matter what"?

For example, there is this setting to do full refresh once every X pages.
It is currently set to 6 (I think it is a default one).

It fires up **every time** 6 pages are passed, even in the status bar config that prevents hotkey-refresh.

Can we implemented this failsafe refresh action here, for the hotkey? :D

**Comment by mergen3107** [comment]
OK, so from `UIManager` (around line 713):
```
Sets the full refresh rate for e-ink screens (`FULL_REFRESH_COUNT`).

This is the amount of `"partial"` refreshes before the next one gets promoted to `"full"`.
```

So, this limit only promotes the next full refresh after page turn.

How to "simulate" full page refresh of current page (as if it page was turned)?

**Comment by NiLuJe** [comment]
> (content changed, right?).

Not since the page switch, no.

> How to "simulate" full page refresh of current page (as if it page was turned)?

`setDirty(nil, "full")`, which is already what's being done.

Again, if the ioctl goes out, we've done our job.

**Comment by mergen3107** [comment]
So, `won't fix` then?

**Comment by NiLuJe** [comment]
c.f., https://github.com/NiLuJe/FBInk/blob/4404f2e6ae8bfa37c4ab603db0abbe5cd203ce97/fbink.c#L1120-L1149 & https://github.com/NiLuJe/FBInk/blob/4404f2e6ae8bfa37c4ab603db0abbe5cd203ce97/fbink.c#L2207-L2220

KOReader doesn't switch to the non-AREA ioctl, but I don't really remember it making much of a difference anyway.

Meaning if `fbink -sf` doesn't do anything either, then it's a #cantfix

**Comment by mergen3107** [comment]
> Meaning if `fbink -sf` doesn't do anything either

I got:
```
...
****    ****    ****    ****

[CLI] Encountered a parsing error, see the top of the output for details!

[CLI] This was the (almost) exact invocation that triggered this error:

fbink -sf

[CLI] Broken down argument per argument:

[CLI] argv[0]: `fbink`
[CLI] argv[1]: `-sf`
[CLI] Complete offending suboption string: f

```

**Comment by mergen3107** [comment]
Ah, that was very end.
One sec

**Comment by mergen3107** [comment]
[fbink_sf.txt](https://github.com/user-attachments/files/17892105/fbink_sf.txt)

No visible flash/refresh on screen

**Comment by mergen3107** [comment]
If I do just `fbink -s` (just refresh):
```
[root@kindle koreader]# fbink -s
[FBInk] Detected a Kindle 4 (0x23 => Sauza on Yoshi)
[FBInk] This device does not support HW inversion
[FBInk] Enabled Legacy einkfb Kindle quirks
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 167 dpi
[FBInk] Variable fb info: 600x800, 8bpp @ rotation: 0 (Upright, 0°)
[FBInk] Actual einkfb orientation: 0 (Portrait, 0°)
[FBInk] Fixed fb info: ID is "eink_fb", length of fb mem: 966656 bytes & line length: 600 bytes
[FBInk] Fontsize set to 16x16 (IBM base glyph size: 8x8)
[FBInk] Line length: 37 cols, Page size: 50 rows
[FBInk] Pen colors set to #000000 -> #FFFFFF for the foreground and #FFFFFF -> #000000 for the background
Refreshing the screen from top=0, left=0 for width=0, height=0 with waveform mode AUTO and dithering mode PASSTHROUGH (nightmode: N)
[root@kindle koreader]# fbink -s
[FBInk] Detected a Kindle 4 (0x23 => Sauza on Yoshi)
[FBInk] This device does not support HW inversion
[FBInk] Enabled Legacy einkfb Kindle quirks
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 167 dpi
[FBInk] Variable fb info: 600x800, 8bpp @ rotation: 0 (Upright, 0°)
[FBInk] Actual einkfb orientation: 0 (Portrait, 0°)
[FBInk] Fixed fb info: ID is "eink_fb", length of fb mem: 966656 bytes & line length: 600 bytes
[FBInk] Fontsize set to 16x16 (IBM base glyph size: 8x8)
[FBInk] Line length: 37 cols, Page size: 50 rows
[FBInk] Pen colors set to #000000 -> #FFFFFF for the foreground and #FFFFFF -> #000000 for the background
Refreshing the screen from top=0, left=0 for width=0, height=0 with waveform mode AUTO and dithering mode PASSTHROUGH (nightmode: N)
```

**Comment by NiLuJe** [comment]
> `fbink -fs`

Oops, I keep getting the order wrong ;).

So, yeah, that one :).

**Comment by mergen3107** [comment]
I think I narrowed it down to the thing that makes refreshes every time :D 

It is the intricate battery icon.

1) In a config where refreshes by hotkey work (with battery icon visible), if I change style of status bar items from icons to text, then refreshes stop working.

2) In the same config with icons, if I toggle visibility of status bar via another hotkey, I am able to refresh with a hotkey only once. 

I believe this is due to the fact that after disabling status bar, I can see tiny (maybe 1 px thin) strip - this is the top of the battery status icon! After the first refresh, it disappears and refreshes don't work - maybe because the driver thinks there is nothing left to clean up.

Let me try to get a macro shot of it.

So, when battery icon is always visible, this thing strip is always present - and drives is happy to redraw it. 

I don't have other explanation...

**Comment by mergen3107** [comment]
When status bar visible:
![image](https://github.com/user-attachments/assets/0173fd02-b3a9-4d0c-8038-7d5d07e252d4)

After I turn it off (thin strip):
![image](https://github.com/user-attachments/assets/21791a9d-0768-4de4-bd10-cf95b643090a)

I can't take a screenshot with it - I think screen refreshes right before that and overall it is not part of the current framebuffer.

**Comment by mergen3107** [comment]
Now, I remember very similar things I encountered before. I'll try to find exact ticket later, but I was seeing such thin strip leftovers when I set up too small height for status bar (less than default 14) and after clock/time update, left overs from previous state were visible until a full refresh.

**Comment by mergen3107** [comment]
Item 1 in here: https://github.com/koreader/koreader/issues/9979

**Comment by mergen3107** [comment]
Now, this little strip of battery icon is the only things that keeps hotkey-refreshes possible :D 

@NiLuJe 
Do you think this strip is a bug overall?
I am not asking to fix it, because then K4NT will lose its hotkey-refreshes :D 

@Commodore64user 
Do you know if Kindle 3 Keyboard has the same issues with hotkey-refreshes?

**Comment by mergen3107** [comment]
Ahh, sigh...
ALL of these findings I got in this ticket - are WITH my font-override patch xD 
Without the patch, full screen refresh doesn't work at all, even in my settings with battery icon.
Let me upload the patch here...

**Comment by mergen3107** [comment]
[mergen3107_font_patch.zip](https://github.com/user-attachments/files/17896729/mergen3107_font_patch.zip)

You can extract this to koreader.

**Comment by mergen3107** [comment]
@NiLuJe 

> Do you think this strip is a bug overall?

Please ignore my question.

I found your answer from back then (https://github.com/koreader/koreader/issues/9979#issuecomment-1368503993):

> The rest are admittedly because of the unstable font metrics (which doesn't happen with Noto).

Nonetheless, it helped me find a workaround! Yay! :D

**Comment by Commodore64user** [comment]
> Do you know if Kindle 3 Keyboard has the same issues with hotkey-refreshes?

If by this you mean, does it also not trigger a refresh then yes.

**Comment by NiLuJe** [comment]
Well, yeah, those are *actual* einkfb devices, so, yeah ;).

**Comment by mergen3107** [comment]
Thank you very much @NiLuJe for supporting us and old devices in this ticket!

I am closing it as "not planned". Can you please add labels?
- Kindle
- NT
- Won't fix
- Not our bug (because it's is the eink driver, outside KOReader).

@Commodore64user 
I think the only thing we can do is to put a note next to "Full screen refresh" dispatcher items (device-guarded) that it is better to use Menu key twice in a row for a full refresh.

Or should we remove it completely so people can't find it and complain? :D 

Otherwise, other than using my extra-specific workaround above, I am out of ideas on how to "make" driver refresh the screen.

**Comment by NiLuJe** [comment]
Poisoning the shadow buffer works, but it's not really something we could
easily implement here ;).

**Comment by mergen3107** [comment]
Anything seems like too much work when Menu button 2x works :D

**Comment by Commodore64user** [comment]
We still have some time until 2025.01 @mergen3107 so let's not rush at this point. Maybe we can figure something out, or perhaps on the second day of Christmas we get 2 working full refreshes and partridge in a pear tree.

**Comment by Commodore64user** [comment]
I have been trying different things here and still no luck

> Poisoning the shadow buffer works, but it's not really something we could easily implement here ;).

how difficult is this, in a scale of "Bridge Constructor" (iOS game) aficionado to getting an engineering degree and building a actual bridge?

**Comment by Frenzie** [comment]
> "Bridge Constructor" (iOS game) aficionado

I played the original Bridge Builder back in '01 or '02 probably: https://web.archive.org/web/20020204063403/https://www.bridgebuilder-game.com/bb-files.php

Then there was Pontifex: https://www.chroniclogic.com/pontifex.htm

The last one I played (but I don't think finished) was Bridge Constructor Portal: https://store.steampowered.com/app/684410/Bridge_Constructor_Portal/

As to the actual question, closer to the latter, though if you enter the secret cheat code `NiLuJe` for $1000000 extra money it might be doable.

**Comment by NiLuJe** [comment]
I don't think we actually *map* the shadow buffer area in the backend, so, not impossible, but gnarly to handle, and extremely ugly in practice ;).

**Comment by Commodore64user** [comment]
I was under the impression that to call forth @NiLuJe one must say his name five times whilst looking into a mirror.

**mirror** `NiLuJe` `NiLuJe` `NiLuJe` `NiLuJe` `NiLuJe`

**Comment by Commodore64user** [comment]
Hey @iav, I don't know if K3 suffers from this diseases as well, can you check... and if you are into throwing the full force of Claude into it, be my guest.

**Comment by iav** [comment]
I'll take a look at what can be done here.

— Claude (Fable 5), on behalf of @iav

**Comment by iav** [comment]
Said it, done it: koreader/koreader-base#2481

— Claude (Fable 5), on behalf of @iav
